.github/workflows/testing.yml

```diff
@@ -171,6 +172,13 @@ jobs:
         run: |
           docker rm taichi_build taichi_test -f
 
+      - name: Save wheel if test failed
+        if: failure() && steps.test.conclusion == 'failure'
```

Comment:
Seems the github syntax says `${{ failure() && steps.demo.conclusion == 'failure' }}` ? https://docs.github.com/en/actions/learn-github-actions/expressions#example-10
are these the same? ;/

Reply:
It says

> When you use expressions in an if conditional, you may omit the expression syntax 

in the same page :)

Reply:
cool, didn't know that

Reply:
What is the difference between `failure()` and `steps.demo.conclusion == 'failure'`?

Reply:
> What is the difference between `failure()` and `steps.demo.conclusion == 'failure'`? Or does one imply the other?

as far as i understand from the doc (link above), the first part is true as long as any previous steps fail, while the second part specifies a condition for that failure, namely `steps.test.conclusion` the test (id) part should fail, not the build step for example.

Reply:
Fine, as the example uses that expression.